From d5bb740133388e4470e6508f5ff7ea3fb8103a25 Mon Sep 17 00:00:00 2001
From: Rune Lillesveen <futhark@chromium.org>
Date: Thu, 10 Oct 2024 12:31:14 +0000
Subject: [PATCH] Call RebuildLayoutTreeForChild for pseudo elements

Use the same code path as for normal elements to make sure the
WhiteSpaceAttacher marks pseudo elements as visited to get correct
whitespace handling. Also use RebuildPseudoElementLayoutTree() for the
::marker pseudo since the implementation was basically the same.

Fixes an incorrect whitespace removal bug and a fuzzer crash caused by
inconsistent ::first-letter texts when size containers were involved.

Bug: 372252513, 372358471
Change-Id: I1f99f3b83809b9eef72b42aec9ad099bf3c56ded
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5920416
Commit-Queue: Rune Lillesveen <futhark@chromium.org>
Reviewed-by: Morten Stenshorne <mstensho@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1366747}
---

--- a/third_party/blink/renderer/core/dom/element.cc
+++ b/third_party/blink/renderer/core/dom/element.cc
@@ -4408,7 +4408,7 @@
       RebuildChildrenLayoutTrees(*child_attacher);
     }
     RebuildPseudoElementLayoutTree(kPseudoIdBefore, *child_attacher);
-    RebuildMarkerLayoutTree(*child_attacher);
+    RebuildPseudoElementLayoutTree(kPseudoIdMarker, *child_attacher);
     RebuildPseudoElementLayoutTree(kPseudoIdScrollMarkerGroupBefore,
                                    local_attacher);
     RebuildPseudoElementLayoutTree(kPseudoIdScrollPrevButton, *child_attacher);
@@ -4435,9 +4435,7 @@
     PseudoId pseudo_id,
     WhitespaceAttacher& whitespace_attacher) {
   if (PseudoElement* element = GetPseudoElement(pseudo_id)) {
-    if (element->NeedsRebuildLayoutTree(whitespace_attacher)) {
-      element->RebuildLayoutTree(whitespace_attacher);
-    }
+    RebuildLayoutTreeForChild(element, whitespace_attacher);
   }
 }
 
@@ -4467,14 +4465,6 @@
     }
   }
 }
-
-void Element::RebuildMarkerLayoutTree(WhitespaceAttacher& whitespace_attacher) {
-  if (PseudoElement* marker = GetPseudoElement(kPseudoIdMarker)) {
-    if (marker->NeedsRebuildLayoutTree(whitespace_attacher)) {
-      marker->RebuildLayoutTree(whitespace_attacher);
-    }
-  }
-}
 
 void Element::HandleSubtreeModifications() {
   if (auto* layout_object = GetLayoutObject()) {
--- a/third_party/blink/renderer/core/dom/element.h
+++ b/third_party/blink/renderer/core/dom/element.h
@@ -752,7 +752,7 @@
   void RebuildLayoutTreeForTraversalRootAncestor() {
     RebuildFirstLetterLayoutTree();
     WhitespaceAttacher whitespace_attacher;
-    RebuildMarkerLayoutTree(whitespace_attacher);
+    RebuildPseudoElementLayoutTree(kPseudoIdMarker, whitespace_attacher);
     HandleSubtreeModifications();
   }
   void RebuildLayoutTreeForSizeContainerAncestor() {
@@ -1708,7 +1708,6 @@
 
   void RebuildPseudoElementLayoutTree(PseudoId, WhitespaceAttacher&);
   void RebuildFirstLetterLayoutTree();
-  void RebuildMarkerLayoutTree(WhitespaceAttacher&);
   void RebuildShadowRootLayoutTree(WhitespaceAttacher&);
   inline void CheckForEmptyStyleChange(const Node* node_before_change,
                                        const Node* node_after_change);
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/css/css-conditional/container-queries/crashtests/chrome-bug-372358471-crash.html
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
+<title>::first-letter text crash during size query change</title>
+<link rel="help" href="https://crbug/com/372358471">
+<style>
+  body {
+    &::first-letter { color: green; }
+    &::before { content: "."; }
+  }
+  canvas {
+    float: left;
+    container-type: inline-size;
+    &::first-line {
+      color: pink;
+    }
+  }
+  :first-of-type {}
+</style>
+<body> <canvas></canvas><span id="id_1"></span><span>x</span>
+<script>
+  document.body.offsetTop;
+  id_1.remove();
+  document.body.offsetTop;
+</script>
+
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/css/css-pseudo/before-preceding-whitespace-dynamic.html
@@ -0,0 +1,15 @@
+<!DOCTYPE html>
+<title>Removing an empty inline should not cause a space after ::before content to collapse</title>
+<link rel="help" href="https://drafts.csswg.org/css-pseudo/#generated-content">
+<link rel="help" href="https://drafts.csswg.org/css-text/#white-space-rules">
+<link rel="match" href="../reference/pass_if_two_words.html">
+<style>
+  div::before { content: "two" }
+</style>
+<p>There should be a space between "two" and "words" below.</p>
+<div> <span id="rm"></span><span>words</span></div>
+<script>
+  rm.offsetTop;
+  rm.remove();
+  rm.offsetTop;
+</script>
